plugin/pkg/auth/authorizer/node/graph.go

```diff
@@ -0,0 +1,244 @@
+}
+
+func NewNode(nodeType nodeType, namespace, name string, id int) *NamedNode {
+	return &NamedNode{
+		nodeType:  nodeType,
+		name:      name,
+		namespace: namespace,
+		id:        id,
+	}
+}
+func (n *NamedNode) ID() int {
+	return n.id
+}
+func (n *NamedNode) String() string {
+	if len(n.namespace) == 0 {
+		return nodeTypes[n.nodeType] + ":" + n.name
+	}
+	return nodeTypes[n.nodeType] + ":" + n.namespace + "/" + n.name
+}
+
+// NodeEdge is a graph edge that is related to a kubernetes Node
+type NodeEdge struct {
+	F    graph.Node
+	T    graph.Node
+	Node graph.Node
+}
+
+func NewNodeEdge(from, to, node graph.Node) graph.Edge {
+	return &NodeEdge{F: from, T: to, Node: node}
+}
+func (e *NodeEdge) From() graph.Node { return e.F }
+func (e *NodeEdge) To() graph.Node   { return e.T }
+func (e *NodeEdge) Weight() float64  { return 0 }
+func (e *NodeEdge) NodeID() int      { return e.Node.ID() }
+
+// Graph holds graph nodes and a way to look up a node for a particular API type/namespace/name
+type Graph struct {
+	lock  sync.RWMutex
+	graph *simple.DirectedAcyclicGraph
+	nodes map[nodeType]map[string]map[string]*NamedNode
+}
+
+func NewGraph() *Graph {
+	return &Graph{
+		nodes: map[nodeType]map[string]map[string]*NamedNode{},
+		graph: simple.NewDirectedAcyclicGraph(0, 0),
+	}
+}
+
+type nodeType byte
+
+const (
+	ConfigMapNode nodeType = iota
+	NodeNode
+	PodNode
+	PVCNode
+	PVNode
+	SecretNode
+)
+
+var nodeTypes = map[nodeType]string{
+	ConfigMapNode: "configmap",
+	NodeNode:      "node",
+	PodNode:       "pod",
+	PVCNode:       "pvc",
+	PVNode:        "pv",
+	SecretNode:    "secret",
+}
+
+// must be called under a write lock
+func (g *Graph) getOrCreateNode(nodeType nodeType, namespace, name string) *NamedNode {
+	if node, exists := g.getNode(nodeType, namespace, name); exists {
+		return node
+	}
+	return g.createNode(nodeType, namespace, name)
+}
+
+// must be called under a read lock
+func (g *Graph) getNode(nodeType nodeType, namespace, name string) (*NamedNode, bool) {
+	node, exists := g.nodes[nodeType][namespace][name]
+	return node, exists
+}
+
+// must be called under a write lock
+func (g *Graph) createNode(nodeType nodeType, namespace, name string) *NamedNode {
+	typedNodes, exists := g.nodes[nodeType]
+	if !exists {
+		typedNodes = map[string]map[string]*NamedNode{}
+		g.nodes[nodeType] = typedNodes
+	}
+
+	namespacedNodes, exists := typedNodes[namespace]
+	if !exists {
+		namespacedNodes = map[string]*NamedNode{}
+		typedNodes[namespace] = namespacedNodes
+	}
+
+	node := NewNode(nodeType, namespace, name, g.graph.NewNodeID())
+	namespacedNodes[name] = node
+	g.graph.AddNode(node)
+
+	return node
+}
+
+// must be called under write lock
+func (g *Graph) deleteNode(nodeType nodeType, namespace, name string) {
+	node, exists := g.getNode(nodeType, namespace, name)
+	if !exists {
+		return
+	}
+
+	// find existing neighbors
+	neighborsToRemove := []graph.Node{}
+	g.graph.VisitFrom(node, func(neighbor graph.Node) bool {
```

Comment:
my addition (see `gonum: visiting graph traversal` commit) to save allocating a slice of all neighbors for every step of every read traversal (we really need reads to be cheap and fast). saves ~95% on allocations for highly shared secrets